### PR TITLE
[WIP] Bootstrap v4

### PR DESCRIPTION
**Pull request by mattbrictson**
This is our official PR that consolidates all changes needed to support Bootstrap v4.

**This PR is a work in progress.** It may or may not be merged, depending on whether we decide to publish this as a separate version or even spin it off as a separate gem. If you have ideas or feedback on these logistical topics, please join the discussion here: https://github.com/bootstrap-ruby/rails-bootstrap-forms/issues/268

**If you would like to try this preliminary Bootstrap v4 support in your Rails project, use the following Gemfile entry:**

```ruby
gem "bootstrap_form", git: "https://github.com/bootstrap-ruby/rails-bootstrap-forms.git", branch: "bootstrap-v4"
```

**If you would like to contribute to this branch (i.e. fix bugs or provide wrappers for additional v4 features), please open your own PR and use `rails-bootstrap-forms:bootstrap-v4` as the base branch.** If and when your PR is reviewed and merged, it will become part of this larger PR.

Thanks for your help!

### CODE REVIEW

**Comment by GBH** [comment]
This PR probably should be closed as it's not correct for Boostrap4 beta3

- `has_danger` is not a thing: https://getbootstrap.com/docs/4.0/components/forms/#server-side
- `form-control-feedback` not a thing either. It's `invalid-feedback`
- `form-control-label` not a thing. `col-form-label` is used for horizontal forms though: https://getbootstrap.com/docs/4.0/components/forms/#horizontal-form

Let's close this and promote bootstrap-v4 to master so we can start working on it.

**Comment by mattbrictson** [comment]
@GBH good finds. Should we still merge this branch into `master` as originally planned, and then address these issues from there?

**Comment by mattbrictson** [comment]
Rebased against latest `master`

**Comment by GBH** [comment]
@mattbrictson I think it's fine to merge. Let's just promote everything v4 to master so new PRs can be made.

**Comment by mattbrictson** [comment]
@GBH great, thanks! 👍 

@lcreid, @donv, @desheikh: any concerns before I merge this?

File: README.md [hunk 1]
```diff
@@ -468,10 +468,10 @@ error will be displayed below the field. Rails normally wraps the fields in a
 div (field_with_errors), but this behavior is suppressed. Here's an example:
 
 ```html
-<div class="form-group has-error">
-  <label class="control-label" for="user_email">Email</label>
+<div class="form-group has-danger">
```
Comment (lcreid) [inline README.md]: As @GBH pointed out, `has-danger` isn't in Bootstrap 4 anymore. We should either fix the PR, or merge and remember to fix. Whatever is easier.

File: lib/bootstrap_form/form_builder.rb [hunk 2]
```diff
@@ -263,11 +265,11 @@ def control_class
     end
 
     def label_class
-      "control-label"
+      "form-control-label"
     end
 
     def error_class
-      "has-error"
+      "has-danger"
```
Comment (lcreid) [inline lib/bootstrap_form/form_builder.rb]: TODO: Fix this class. I think we want `is-invalid` now, but correct me if I'm wrong.

File: README.md (hunk 1 above)
Comment (GBH) [reply README.md]: I vote for merge and fix later or this is going to be sitting here for weeks.

**Comment by mattbrictson** [comment]
Thanks @lcreid, I added a reminder to #361 to deal with `has_danger`, `form-control-feedback`, and `form-control-label`. I'll merge this now.

**Pull request merged**